### ISSUE TITLE
Offer a lightweight Elastic Agent

### ISSUE THREAD

**Issue by jamesagarside** [elastic/elastic-agent#3364]
When running Elastic Agent as a Daemon Set on Kubernetes it can soon become a rather heavy resource consumer within the cluster. It would be great if we offered a minimal image for situations like this and have Agent pull assets from an EPR. 
This is currently blocking one of my customers adopting Elastic agent. 

**Describe a specific use case for the enhancement or feature:**
If a customer with 100 nodes of Kubernetes want to monitor metrics from each node they are required to run Agent as a daemon set which contains all Beats when actually all they require is Metricbeat. This not only consumed memory but storage also. 

Ensure the Elastic Agent container can ship with the lightest of resources and pull and further assets when required. 

UPDATE: Changed title to widen the scope of this issue. VMs, Server, Edge devices and containerised systems should all be considered when looking in to this issue.

**Comment by jlind23** [comment]
Bring @nimarezainia here, probably worth having a discussion on this as we have a lot of ongoing initiative related to agent on k8s.

**Comment by nimarezainia** [comment]
@jamesagarside we are thinking about modifying the [agent footprint (roadmap)](https://github.com/elastic/ingest-dev/issues/2193) that would not include all the integrations, which has bloated the current agent. However, removing Beats and only adding them to the agent when needed is not a consideration.

Do you know what the ideal image size would be in this case? what are the integrations they are using?

**Comment by jamesagarside** [comment]
Hi @nimarezainia and @jlind23 

Thank you both for coming back to me on this and thank you for linking the Agent Footprint roadmap item. I see the acceptance criteria is for the image to be 500MB, this is still pretty large is it not? 
Would it be possible to explore stripping back to just Elastic Agent and pulling the required binaries when needed? 
I believe this requirement will become more and more prevalent as we build out more solution focused offerings with server-less. An example would be customers looking to use O11Y might only care about Metrics and therefore would be sitting with redundant binaries such as Auditbeat, Cloudbeat, Endpoint Security, Filebeat, OSQuerybeat & Packetbeat when they really might only care about Metricbeat and Heartbeat. This would not only shrink the size of the image while reducing resource consumption but would also shrink the toolbox for living off the land attacks.

**Comment by jerrac** [comment]
@nimarezainia FYI, the roadmap link 404's.

I'd like to suggest this not be just for Kube or containers. Would it make sense for @jamesagarside to modify the title and initial post to generalize this issue into a request for a smaller Agent overall? Or should I start a new issue?

Yesterday I was trying to install Agent on some vm's, and it failed due to needing over 1.7G of disk, and all the servers having small disk and partition sizes.

I ended up starting this topic: https://discuss.elastic.co/t/are-there-any-issues-or-topics-on-shrinking-elastic-agents-install-size/350099 which led me here.

### Downloading on demand
> However, removing Beats and only adding them to the agent when needed is not a consideration.

I've seen the reasoning for including everything due to iffy networks making downloading the beats chancy, and I agree with it. 

But I do think it would be worth it to offer a version of Agent that does does download on demand for those of us fighting disk space issues. Yes, that adds some more complexity to build and deployment for Elastic, but if we don't have a smaller Agent, we won't be able to use it (and might not be able to use Elastic period.) 

(Side note: To help with the iffy networking issue, would it make sense to have the Fleet server agent download the various binaries, and then the deployed Agents would download those binaries from the Fleet server?)

### Background
In my workplace, our number 1 resource constraint is disk space. We HAVE to make our vm disks tiny. My main VM template is 25G. And I have to have TSDS's for metrics to save on space, which means Agent is a must since [Metricbeat cannot easily use TSDS's](https://discuss.elastic.co/t/anyone-have-an-easy-way-to-make-metricbeat-use-time-series-data-streams-tsds/342375?u=jerrac).

### Other ideas
The only other option I can think of for shrinking the footprint would be make the various beats binaries smaller. But I have a feeling that that would require a very very large amount of development.

**Comment by leandrojmp** [comment]
I also think that this issue should be renamed to deal with the requirement of a lightweight agent not just for containers, but for normal machines and VMs as well.

I have a similar issue, the Agent is too bloated, it requires 1.7 GB after install and may need more than twice this space for installation since you need to download the compressed file, unpack it and run the install command to copy the files to the destination path.

We are planning to start deploying the Elastic Agent on our servers next year and almost all of them have small OS disks, the average OS disk size is 10 GB and the Agent would take almost 20% of it after install and almost 50% during installation, which may not be possible.

Increasing the disk size of thousands of hosts will be add a lot of attrition with the many teams involved and will also need a lot of extra space, which will be in most of the cases unused, it may be pretty hard to justify the Elastic Agent when all other alternatives are way way smaller.

The size until version 8.5 was already big compared with some alternatives, but it was still acceptable, 1.7 GB is too big for the majority of scenarios where you just need a simple log or metric collector.

**Comment by nimarezainia** [comment]
@jerrac & @leandrojmp thank for your comments here. This is a high priority item for us to address and we are having those technical trade-off discussions. We will use this generic issue as the public facing issue. All the options mentioned above are being considered.

**Comment by leandrojmp** [comment]
@nimarezainia good to know that this is a priority item for Elastic.

If it helps, I can provide more context about my use case and issues that we have with the agent being that big.

**Comment by mbudge** [comment]
See this discussion 

https://github.com/elastic/elastic-agent/issues/2759

We've had to repartition a number of linux servers because there wasn't enough disk space to install/upgrade elastic-agent. 

The odd server is ok while elastic-agent is maturing but this needs to be addressed.

These problems can cause system stability issues on linux

- All the beats agents, endpoint, osquery and fleet server are stored in the components folder even though most aren't required.
- Agent log files are stored in the /opt folder

/opt is not meant to store log files. Log files should go in /log/Elastic which should still be captured by diagnostics.

Why not have a separate download for a Fleet server installation of Elastic-Agent. Elastic-Agent connects to Fleet server to download the binary files in the Fleet servers components directory when required. This could be done from on-prem or elastic-cloud Fleet server. Cuts out connections to artefacts.elastic.co and on-prem package repo.

Thanks

**Comment by leandrojmp** [comment]
> This could be done from on-prem or elastic-cloud Fleet server. Cuts out connections to artefacts.elastic.co and on-prem package repo.

Just to point out that this is already possible if you configure it to work as it is in an [air-gapped environment](https://www.elastic.co/guide/en/fleet/current/air-gapped.html#air-gapped).

**Comment by jerrac** [comment]
My concern is that if you try to shove all the monitoring code needed into Agent's binary, you'll end up with the same problem we have now. I mean, each beat is many many many times larger now than it was a few years ago. That, I'm guessing, is due to all the new features added. So combining all those features into one is going to result is something a lot larger. Even if you get some savings by not duplicating the shared parts of the beats.

For context, I've been trying to use ES since it was 0.26.0 and logstash-forwarder was a tiny little thing. So when I figured out that the beats had morphed into 100+MB files, I was rather surprised. (And disappointed.)

I'm also very leery of the way features between beats and agent are not kept in sync. A while back I was trying to get around some of the issues with Agent by just deploying beats, and ran into the fact that metricbeat doesn't know how to autoconfigure TSDS's. I also ran into the fact that Agent doesn't support docker autodiscovery hints. So I've been trying to get my stuff monitored by running both Filebeat (for the autodiscovery of docker logs), and Agent (for the space savings from TSDS's.)

So if you shove everything into Agent, I expect the beats will get abandoned, and we won't actually see the resource savings this topic is about.

Anyway, those concerns aside, I do have a couple more ideas.

One strategy that would partially meet the needs of this issue would be to ask administrators to install (and upgrade) the needed beats outside of Agent. In my case I install via the apt/yum repos, so I really don't need Agent to download the beats. Then we let Agent manage the beat's configuration and do all the extra integration stuff it does. 

I'd also suggest not running a beats process per integration on the policy. I've configured two system integrations, one is just monitoring a list of specific processes, while the other monitors everything else. That means I have 2 (at least) instances of metricbeat running. I think I ended up with at least 1 other for the docker metrics integration...

**Comment by mbudge** [comment]
> I'm also very leery of the way features between beats and agent are not kept in sync. A while back I was trying to get around some of the issues with Agent by just deploying beats, and ran into the fact that metricbeat doesn't know how to autoconfigure TSDS's. I also ran into the fact that Agent doesn't support docker autodiscovery hints. So I've been trying to get my stuff monitored by running both Filebeat (for the autodiscovery of docker logs), and Agent (for the space savings from TSDS's.)

Off topic but if the metricbeat data is delayed and the TSDB index rolls over into a read only ILM phase, elastic returns a 409 forbidden error when trying to index the late data. Logsash keeps trying to index the late events until it gets stuck. Logstash doesn't crash, it just stops processing new log data which causes an outage. They need to add a filter to drop late metrics data either in the logstash elasticsearch plugin or metrics ingest pipelines. https://github.com/elastic/logstash/issues/15832

Back to lightweight agent, can't elastic-agent check if the beats agents exists in the component folder. If it doesn't exist, download it from the public or private package repo then restart the elastic-agent process. On an upgrade, pause log collection, upgrade agent then download and replace the required beats then restart log collection.

Manually copying beats to endpoints might be ok on small networks, but on 1000's of endpoints it would get cumbersome.

**Comment by cmacknz** [comment]
>  If it doesn't exist, download it from the public or private package repo then restart the elastic-agent process. On an upgrade, pause log collection, upgrade agent then download and replace the required beats then restart log collection.

We used to do this prior to 8.6 for the less often used binaries (auditbeat, packetbeat, etc) and saw a significant support load from the download failing. Across our entire user base, things like misbehaving proxies and temporary DNS errors are not uncommon. So we eliminated this entire class of error from the system, this came at the cost of an increased on disk footprint. The on disk footprint has grown a lot since we decided to do this, things like the addition of universal profiling via an acquisition hadn't happened yet when we decided to do this for one example.

We are hesitant to bring back another source of network errors, so our preference is to explore other ways to reduce the disk footprint first. We have initial experiments showing we can shrink the on disk footprint on Linux by 40-50% of what it is today without bringing back on demand downloads. We know this has become a significant pain point.

**Comment by mbudge** [comment]
artefacts.elastic.co should have a static IP which is used when the domain can’t be used. This could be because there is a network issue or a proxy isn’t available.

Those proxy issues are most likely SSL inspection. We had the same problem. SSL inspection was causing a certificate error when downloading from artefacts as the proxy couldn’t validate the root certificate. It was fixed by adding the domain to the SSL Inspection bypass list. Also added to content filtering bypass.

**Comment by Kusig** [comment]
We have about 5000 servers with disk size problems due to the deployment of ElasticAgent ....  
Having a tiny agent with download by need is by far the best and most probably the only solution to this. Please Elastic, you could do better.

**Comment by mbudge** [comment]
5000 endpoints * 1.8GB = 9TB lol

I remember someone saying they added the Beats/Universal Profiling executables to Elastic-Agent because

- Proxy issues
- Problems downloading executables onto Kubernetes nodes

Proxy issues will be SSL inspection and content filtering policies, as customers can do rolling upgrades to take the load off the proxies. 

I think they need to offer a lightweight agent with the core Beats files compressed which can do on-demand download executables which aren't used for the core monitoring functions. 

Another option could be different versions of elastic-agent for cloud, windows and linux endpoints and Kubernetes.

Universal profiling only works on linux so it's kinda pointless deploying it across a windows estate.

**Comment by jlind23** [comment]
@mbudge there is an issue that we are currently adressing here - https://github.com/elastic/beats/pull/38183
Anyway reintroducing on-demand binary download is not part of it as we had many problems with it before.

**Comment by strawgate** [comment]
> Having a tiny agent with download by need is by far the best and most probably the only solution to this. Please Elastic, you could do better.

The Agent always needs Metricbeat and Filebeat which limits the potential footprint reduction of download-on-demand as Metricbeat and Filebeat are the largest beats. We have an alternative solution we are working through that we believe will reduce the disk space requirements about as effectively as download-on-demand while still achieving our goals around installation/upgrade reliability.

> Universal profiling only works on linux so it's kinda pointless deploying it across a windows estate.

The Universal Profiling components are not present on Windows installations or in the Windows installer.

**Comment by legoguy1000** [comment]
> > Having a tiny agent with download by need is by far the best and most probably the only solution to this. Please Elastic, you could do better.
> 
> The Agent always needs Metricbeat and Filebeat which limits the potential footprint reduction of download-on-demand as Metricbeat and Filebeat are the largest beats. We have an alternative solution we are working through that we believe will reduce the disk space requirements about as effectively as download-on-demand while still achieving our goals around installation/upgrade reliability.
> 
> > Universal profiling only works on linux so it's kinda pointless deploying it across a windows estate.
> 
> The Universal Profiling components are not present on Windows installations or in the Windows installer.

I think saying always needs filebeat and metricbeat isn't accurate. We tested the agent without metricbeat since we don't care about metrics for our use case and since we turned off all the metrics integrations, it worked just fine. Filebeat I wouldn't argue since almost every integration uses it.

**Comment by strawgate** [comment]
> I think saying always needs filebeat and metricbeat isn't accurate. We tested the agent without metricbeat since we don't care about metrics for our use case and since we turned off all the metrics integrations, it worked just fine. Filebeat I wouldn't argue since almost every integration uses it.

That's fair, however, Agent uses Metricbeat for its internal health reporting to Fleet which is the default configuration. In a "download-on-demand" world, not having Agent internal monitoring via Metricbeat and Filebeat would make it difficult to understand the health of the agent components including elevating failures related to "download-on-demand" and the resulting disabled integrations/policy components.

We believe the solution we've identified provides the best of both worlds with a significantly smaller binary size and a consistent disk and network footprint. Feel free to follow the progress in the issue linked by @jlind23

**Comment by strawgate** [comment]
> * Has there been any thought around `osqueryd`? It's the second largest binary, ~270MB, and if you're not using OSquery, this is a good amount of overhead. I'm not sure how `osqueryd` plays into `osquerybeat`, but not having ~270MB sit around if not needed would be nice. (I'm excluding `osquerybeat` as that seems to be getting addressed as part of `agentbeat`).

osquery does have an bigger footprint on Linux than the other platforms. Our current plan does not have us addressing the size of osqueryd in this step.

> Are the folders, kibana, lenses, and module really needed to be shipped with Elastic Agent? That seems like ~20MB of free disk space.

These are an artifact of us including individual beats so they should not be present in the resulting Agent installation once the agentbeat work is merged. It also will significantly reduce the file count which should have a significant improvement to installation time on systems with slower disks (hard disk drives).

**Comment by cmacknz** [comment]
> Has there been any thought around osqueryd? It's the second largest binary, ~270MB, and if you're not using OSquery, this is a good amount of overhead.

It turns out the osqueryd binary we pull out of the upstream osquery .tar.gz isn't stripped. It looks like this changed at some point looking at the upstream issues and we didn't notice.

https://github.com/elastic/beats/issues/38890

**Comment by santosj1989** [comment]
Summarizing the story, Elastic still doesn't have an alternative for the size of the elastic-agent? I believe this is the biggest problem for all Sysadmins here. Computing parks with hundreds of machines and always being forced to cut costs. 1.7GB per machine is complicated. Has anyone been able to apply the lightweight option of this agent?

**Comment by jerrac** [comment]
@santosj1989 Looks like they're seriously looking to combine all the beats agents into one for Elastic Agent. See https://github.com/elastic/beats/pull/38183

It does seem to save some space.

Still not sure I think it's the best way to do things... 'Course I've never been fond of Agent in the first place, so I'm rather biased.

I haven't tried it yet, but you could try just using collectd if all you need is metrics. I THINK they have a lighter footprint. Logstash (at least) can receive collectd data.

**Comment by pierrehilbert** [comment]
Hello @santosj1989

> Summarizing the story, Elastic still doesn't have an alternative for the size of the elastic-agent? I believe this is the biggest problem for all Sysadmins here. Computing parks with hundreds of machines and always being forced to cut costs. 1.7GB per machine is complicated. Has anyone been able to apply the lightweight option of this agent?

As @nimarezainia mentioned two months ago, we are seeing this topic as a high priority one and are currently working on it.

**Comment by BenB196** [comment]
I see this issue is closed, but for anyone who wants to "live on the edge" (pun somewhat intended), you can "make your own" edge deployment by just remove any binaries you don't want and technically the Elastic Agent will continue to work (at least for now, as I'm sure it's almost definitely not supported by Elastic and it can break at anytime).

For example, if you only care about a true "edge" deployment, you can go into the `components` directory and just remove the non-"edge" stuff; `rm -f apm-server* cloud-defend fleet-server* pf-elastic-collector* pf-elastic-symbolizer*`. That'll save you another ~275MB, but just note that you'd need to reinstall the Elastic Agent if you want to use integrations which require those binaries. (If you don't need it, you can also delete `cloudbeat*` saving another ~280MB if you don't need it.

**Comment by cmacknz** [comment]
The overall size reduction for 8.14 is more than what is in that PR. The extracted size of the Linux .tar.gz will be ~1 GB in 8.14 or ~800 MB smaller than 8.13. The final numbers for all of the operating systems will be in the 8.14 release notes.

1 GB is still pretty big I agree, but that 800 MB reduction is a lot of progress that didn't require us to introduce any compromises. I also think there are still some additional MBs we can shave off but the 8.14.0 release will likely be the biggest reduction in size for some time.

The core problem is that Elastic Agent is capable of doing a lot. It can both collect OS system metrics and implement an Endpoint protection service and many things in between. It is actually quite difficult for us to shrink the existing size without introducing a compromise for somebody's core use case. The biggest con of this strategy is that the disk footprint is always larger than it needs to be unless you are using every single capability of agent.

There are some larger architecture changes we are planning that I hope will continue to bring the size down over time, but those are tied to larger efforts than just shrinking the disk footprint for the sake of it. I can't link to public tracking issue for this quite yet but I hope to be able to soon.

**Comment by strawgate** [comment]
@nimarezainia can we add this and https://github.com/elastic/elastic-agent/issues/4599 to the list of issues we tag when we have a public tracking issue for the architectural changes?

**Comment by leandrojmp** [comment]
@cmacknz a final size of 1GB while still pretty big it seems to be on par with other similar tools, if I'm not wrong the datadog agent has a similar size, something around 1.3 GB.

Not sure what are the changes being planned but as a feedback, I have 2 main use cases for the Elastic Agent.

One is using the Elastic Agent on edge hosts to collect system logs and metrics, on this case the inputs required would be reading log files, windows events, auditd, journald and things like that.

The other use case would be using the Elastic Agent to get logs from API endpoints, Cloud Buckets, Message Brokers, things like that where you would normally use a tool like Logstash and also like Logstash you would just need the agent that collect these things to run in just a couple of servers.

So all the inputs for AWS, GCS, PubSub, EventHubs, BlobStorage, Kafka etc are in most of the case not required or used when you have the agent on edge hosts.

Maybe having 2 types of agents where one would be a slim version without all those inputs not commonly used on edge hosts could help reduce the size?

**Comment by mbudge** [comment]
Another option could be a setting in the Fleet UI which allows users to delete executables which aren’t required. Executables could be deleted at installation/upgrade time.

for example if metricbeat, osquery, universal profiling or elastic defend aren’t required, let users delete the executables through the fleet policy. This could be undone through a change in policy at the next upgrade.

**Comment by strawgate** [comment]
> One is using the Elastic Agent on edge hosts to collect system logs and metrics, on this case the inputs required would be reading log files, windows events, auditd, journald and things like that.

While the exact numbers will be in the release notes, it should be noted that the Windows and macOS footprints are looking like they will both be <600MB in 8.14 (perhaps closer to 500MB than 600MB) -- https://github.com/elastic/ingest-docs/pull/1079/files

We know our work isn't done here and as Craig said, we have plans to do more work in this area.

**Comment by jamesagarside** [comment]
A thought on container distribution. I was listening to the Kubernetes OnGres podcast the other day and they OnGres team talked about how they dynamically package containers to enable all the permutations of plugins to be possible. Its talked about in this blog https://www.ongres.com/blog/why-postgres-extensions-should-be-distributed-and-packaged-as-oci-images/

Could we adopt something similar for customers who require the absolute minimum footprint. It would enable users to select only the packages they need in an image.

**Comment by jerrac** [comment]
Has it been long enough we can go open a new issue? 

For the elastic-agent-8.15.0-linux-x86_64.tar.gz file, you need at least 1.3G to both download and extract.

That is not lightweight.

**Comment by nimarezainia** [comment]
@jerrac I understand your concern, this is a big reduction from where we were and we continue to seek opportunities where this can be reduced further. Right now however we don't have a timeline on when the next reduction in size will occur.
